lib/ansible/module_utils/facts/hardware/linux.py

```diff
@@ -443,51 +446,85 @@ def _mtab_entries(self):
-                continue
+            device, mount, fstype, options = fields[0], fields[1], fields[2], fields[3]
 
-            if fstype == 'none':
+            if not device.startswith('/') and ':/' not in device or fstype == 'none':
                 continue
 
-            mount_statvfs_info = get_mount_size(mount)
-
-            if mount in bind_mounts:
-                # only add if not already there, we might have a plain /etc/mtab
-                if not self.MTAB_BIND_MOUNT_RE.match(options):
-                    options += ",bind"
-
-            # _udevadm_uuid is a fallback for versions of lsblk <= 2.23 that don't have --paths
-            # see _run_lsblk() above
-            # https://github.com/ansible/ansible/issues/36077
-            uuid = uuids.get(device, self._udevadm_uuid(device))
-
             mount_info = {'mount': mount,
                           'device': device,
                           'fstype': fstype,
-                          'options': options,
-                          'uuid': uuid}
+                          'options': options}
 
-            mount_info.update(mount_statvfs_info)
+            if mount in bind_mounts:
+                # only add if not already there, we might have a plain /etc/mtab
+                if not self.MTAB_BIND_MOUNT_RE.match(options):
+                    mount_info['options'] += ",bind"
+
+            results[mount] = {'info': mount_info,
+                              'extra': pool.apply_async(self.get_mount_info, (mount, device, uuids)),
+                              'timelimit': time.time() + maxtime}
+
+        pool.close()  # done with new workers, start gc
+
+        # wait for workers and get results
+        while results:
+            for mount in results:
+                res = results[mount]['extra']
+                if res.ready():
+                    if res.successful():
+                        mount_size, uuid = res.get()
+                        if mount_size:
+                            results[mount]['info'].update(mount_size)
+                        results[mount]['info']['uuid'] = uuid or 'N/A'
+                    else:
+                        # give incomplete data
+                        errmsg = to_text(res.get())
+                        self.module.warn("Error prevented getting extra info for mount %s: %s." % (mount, errmsg))
+                        results[mount]['info']['note'] = 'Could not get extra information: %s.' % (errmsg)
 
-            mounts.append(mount_info)
+                    mounts.append(results[mount]['info'])
+                    del results[mount]
+                    break
+                elif time.time() > results[mount]['timelimit']:
+                    results[mount]['info']['note'] = 'Timed out while attempting to get extra information.'
+                    mounts.append(results[mount]['info'])
+                    del results[mount]
+                    break
```

Comment:
@nitzmahone and I talked about this last night and the timeout is an expectation of how long fact gathering as a whole will take, not how long any specific piece of fact gathering will take.  We thought that probably the best place to enforce that sort of fact gathering would be in the action plugin being worked on in a separate PR.  However, for this PR, it would make more sense to at least try to replicate the status quo as much as possible, ie: take a time at the start of this function.  Then every time we check the time to see if it's time to exit due to timeout, we compare time.time() > function_start_time + timeout to see if it's time to exit.

Reply:
from the docs on setup module
```- gather_timeout
        Set the default timeout in seconds for individual fact gathering```

Reply:
as to what 'individual facts' are we never define, so here i was using the timeout per fact, but running them in parallel as much as possible, the total might take a bit more but it should be close enough. We read all ohter facts serially, but we can improve that with this kind of code going forward.